hal/inc/ble_hal.h

```diff
@@ -15,190 +15,747 @@
-    uint16_t char_count;
-} ble_service;
-
-// Manufacturer-specific data
-typedef struct ble_manuf_data {
-    uint16_t company_id;
-    uint16_t size;
-    const char* data;
-} ble_manuf_data;
-
-// BLE_EVENT_CONNECTED event data
-typedef struct ble_connected_event_data {
-    uint16_t conn_handle;
-} ble_connected_event_data;
+/**
+ * Initialize the BLE stack. This function must be called previous to any other BLE APIs.
+ *
+ * @param[in]   reserved    Reserved for future use.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_stack_init(void* reserved);
 
-// BLE_EVENT_DISCONNECTED event data
-typedef struct ble_disconnected_event_data {
-    uint16_t conn_handle;
-} ble_disconnected_event_data;
+/**
+ * Select the antenna for BLE radio.
+ *
+ * @param[in]   antenna The antenna type, either be BLE_ANT_INTERNAL or BLE_ANT_EXTERNAL.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_select_antenna(hal_ble_ant_type_t antenna);
 
-// BLE_EVENT_CONN_PARAM_CHANGED event data
-typedef struct ble_conn_param_changed_event_data {
-    uint16_t conn_handle;
-} ble_conn_param_changed_event_data;
+/**
+ * Set the callback on BLE events.
+ *
+ * @param[in]   callback    The callback function.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_set_callback_on_events(on_ble_evt_cb_t callback, void* context);
 
-// BLE_EVENT_CHAR_PARAM_CHANGED event data
-typedef struct ble_char_param_changed_event_data {
-    uint16_t conn_handle;
-    uint16_t char_handle;
-} ble_char_param_changed_event_data;
+/**
+ * Set local BLE identity address, which type must be either public or random.
+ *
+ * @param[in]   address Pointer to the BLE local identity address to be set.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_set_device_address(const hal_ble_addr_t* address);
 
-// BLE_EVENT_DATA_SENT event data
-typedef struct ble_data_sent_event_data {
-    uint16_t conn_handle;
-} ble_data_sent_event_data;
+/**
+ * Get local BLE identity address, which type is either public or random static.
+ *
+ * @param[out]  address Pointer to address structure to be filled in.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_get_device_address(hal_ble_addr_t* address);
 
-// BLE_EVENT_DATA_RECEIVED event data
-typedef struct ble_data_received_event_data {
-    uint16_t conn_handle;
-    uint16_t char_handle;
-    const char* data;
-    uint16_t size;
-} ble_data_received_event_data;
+/**
+ * Set the BLE device name.
+ *
+ * @param[in]   device_name Pointer to a UTF-8 encoded, <b>non NULL-terminated</b> string.
+ * @param[in]   len         Length of the UTF-8, <b>non NULL-terminated</b> string pointed to by device_name
+ *                          in octets (must be smaller or equal than @ref BLE_GAP_DEVNAME_MAX_LEN).
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_set_device_name(const uint8_t* device_name, uint16_t len);
```

Comment:
Any reason this cannot simply be `const char* device_name` ? It is a string after all.